system-addon/lib/TopStoriesFeed.jsm

```diff
@@ -79,11 +88,55 @@ this.TopStoriesFeed = class TopStoriesFeed {
       this.storiesLastUpdated = Date.now();
       // This is filtered so an update function can return true to retry on the next run
       this.contentUpdateQueue = this.contentUpdateQueue.filter(update => update());
+
+      this.saveToFile(STORIES_FILE, this.stories);
     } catch (error) {
       Cu.reportError(`Failed to fetch content: ${error.message}`);
     }
   }
 
+  async loadCachedStories() {
+    const {results: stories, timestamp} = await this.loadFromFile(STORIES_FILE);
+    if (stories && stories.length > 0 && this.storiesLastUpdated === 0) {
+      this.dispatchUpdateEvent(this.storiesLastUpdated, {rows: stories});
+      this.storiesLastUpdated = timestamp;
+    }
+  }
+
+  async loadCachedTopics() {
+    const {results: topics, timestamp} = await this.loadFromFile(TOPICS_FILE);
+    if (topics && topics.length > 0 && this.topicsLastUpdated === 0) {
+      this.dispatchUpdateEvent(this.topicsLastUpdated, {topics, read_more_endpoint: this.read_more_endpoint});
+      this.topicsLastUpdated = timestamp;
+    }
+  }
+
+  async loadFromFile(filename) {
```

Comment:
Now that this is getting more involved, I think it makes sense to move to a separate Cache module/component? This would make ```loadFromFile/saveToFile/loadFromPref/saveToPref``` reusable and separate the story feed from the caching concerns.

Wdyt? We can also file a follow-up and refactor later...

Reply:
I think refactoring later might make more sense when we have multiple uses of the cache (although I suppose technically stories + topics = 2 uses of cache…) as that would help provide details of caching requirements, e.g., each caller wants individual `lastUpdated` times, similar `save` frequencies if saving "all cache data" to a single file, etc.

Unless we know now what other things might desire caching. I suppose `LinksCache` could want a local backing of data across restarts vs relying on places. But, in terms of timing of refactoring, I think we would want to think a bit more about `LinksCache` and other possible consumers without blocking this fix.

Reply:
I wonder if dynamic prerendering in #3367 would want to use this disk cache. I think its load behavior is different enough to do something special there, e.g., dynamic prerendering would want the data before messages are sent to content. @k88hudson ?

Reply:
@Mardak @rlr I'd be in favor of creating a Cache.jsm for this, just covering this current functionality (loadFromFile/saveToFile), especially now after adding the Mutex. So that caching functionality and tests are isolated and can be reused and built on. If it's a matter of getting it done, I can help with it tomorrow. We still need the MAX_AGE decision and handle "dismiss" anyway. 

As for other use cases, I will definitely use it for the domain affinity cache which is on my todo list.